src/Components/Server/src/Circuits/CircuitIdFactory.cs

```diff
@@ -2,56 +2,114 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Security.Claims;
 using System.Security.Cryptography;
+using Microsoft.AspNetCore.Components.Server;
 using Microsoft.AspNetCore.DataProtection;
 using Microsoft.AspNetCore.WebUtilities;
-using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Components.Server.Circuits
 {
     // This is a singleton instance
     // Generates strong cryptographic ids for circuits that are protected with authenticated encryption.
     internal class CircuitIdFactory
     {
+        private readonly IDataProtectionProvider _dataProtectorProvider = new EphemeralDataProtectionProvider();
+
         private const string CircuitIdProtectorPurpose = "Microsoft.AspNetCore.Components.Server";
 
         private readonly RandomNumberGenerator _generator = RandomNumberGenerator.Create();
         private readonly IDataProtector _protector;
 
-        public CircuitIdFactory(IDataProtectionProvider provider)
+        public CircuitIdFactory()
         {
-            _protector = provider.CreateProtector(CircuitIdProtectorPurpose);
+            _protector =  _dataProtectorProvider.CreateProtector(CircuitIdProtectorPurpose);
         }
 
         // Generates a circuit id that is produced from a strong cryptographic random number generator
         // we don't care about the underlying payload, other than its uniqueness and the fact that we
         // authenticate encrypt it using data protection.
         // For validation, the fact that we can unprotect the payload is guarantee enough.
-        public string CreateCircuitId()
+        public CircuitId CreateCircuitId()
         {
             var buffer = new byte[32];
             _generator.GetBytes(buffer);
             var payload = _protector.Protect(buffer);
 
-            return Base64UrlTextEncoder.Encode(payload);
+            return new CircuitId(buffer, GetHash(buffer), payload);
+        }
+
+        private static byte[] GetHash(byte[] buffer)
+        {
+            var hashAlgorithm = SHA256.Create();
+            var hash = hashAlgorithm.ComputeHash(buffer);
+            return hash;
+        }
+
+        public bool ValidateCircuitId(string circuitId, string cookie, out CircuitId id)
```

Comment:
Nit: I find the namings of the `circuitId` parameter and the locals in this method hard to follow. It's not obvious to me that `circuitId` should mean "the token from the querystring but not the whole ID" or `rawRequestId` should mean "the bit we got from the querystring" or that `id` should mean "from the cookie". I'd find something like the following easier to interpret at a glance:

```cs
var circuitIdFromCookie = FromCookieValue(cookie);
var requestTokenFromQueryStringBytes = Base64UrlTextEncoder.Decode(requestTokenFromQueryString);
var requestTokenFromCookieBytes = Base64UrlTextEncoder.Decode(circuitIdFromCookie.RequestToken);
if (CryptographicOperations.FixedTimeEquals(requestTokenFromQueryStringBytes, requestTokenFromCookieBytes))
```

Otherwise, interpreting the meanings involves tracing backwards to remember how each data item was derived.